### PR TITLE
CI: Clang CUDA

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

Try to compile with vanilla Clang/LLVM's CUDA C++ frontend & its LLVM PTX backend.

Vanilla LLVM is used in a variety of tools that we want to use for research.

## Additional background

ping @ludgerpaehler

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

File: .github/workflows/cuda.yml [hunk 1]
```diff
@@ -39,6 +39,38 @@ jobs:
 
         cmake --build build -j 2
 
+  # Build libamrex and all tests with CUDA 11.0.2 (recent supported)
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"}
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check -stdlib=libc++"}
```

File: .github/workflows/cuda.yml [hunk 2]
```diff
@@ -39,6 +39,38 @@ jobs:
...
+            -DAMReX_FORTRAN=OFF                          \
+            -DAMReX_PARTICLES=ON                         \
+            -DAMReX_GPU_BACKEND=CUDA                     \
+            -DCMAKE_C_COMPILER=$(which clang)            \
+            -DCMAKE_CXX_COMPILER=$(which clang++)        \
+            -DCMAKE_CUDA_COMPILER=$(which clang++)       \
+            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++)  \
+            -DCMAKE_CUDA_ARCHITECTURES=70                \
+            -DAMReX_CUDA_ARCH=7.0 \
+            -D_CMAKE_CUDA_WHOLE_FLAG="-c"
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
            -DAMReX_CUDA_ARCH=7.0
```

File: .github/workflows/cuda.yml [hunk 3]
```diff
@@ -39,6 +39,38 @@ jobs:
...
+            -DCMAKE_VERBOSE_MAKEFILE=ON                  \
+            -DAMReX_EB=ON                                \
+            -DAMReX_ENABLE_TESTS=ON                      \
+            -DAMReX_FORTRAN=OFF                          \
+            -DAMReX_PARTICLES=ON                         \
+            -DAMReX_GPU_BACKEND=CUDA                     \
+            -DCMAKE_C_COMPILER=$(which clang)            \
+            -DCMAKE_CXX_COMPILER=$(which clang++)        \
+            -DCMAKE_CUDA_COMPILER=$(which clang++)       \
+            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++)  \
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
            -DCMAKE_C_COMPILER=$(which clang-15)            \
            -DCMAKE_CXX_COMPILER=$(which clang++-15)        \
            -DCMAKE_CUDA_COMPILER=$(which clang++-15)       \
            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++-15)  \
```

File: .github/workflows/dependencies/dependencies_nvcc11_ubuntu22.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#
+# License: BSD-3-Clause-LBNL
+
+set -eu -o pipefail
+
+sudo apt-get -qqq update
+sudo apt-get install -y \
+    build-essential     \
+    ca-certificates     \
+    cmake               \
```
Comment (ax3l) [inline .github/workflows/dependencies/dependencies_nvcc11_ubuntu22.sh]:
```suggestion
    clang-15 libc++-15-dev libc++abi-15-dev libc++1-15 libc++abi1-15 \
    cmake               \
```

File: .github/workflows/cuda.yml [hunk 5]
```diff
@@ -39,6 +39,37 @@ jobs:
 
         cmake --build build -j 2
 
+  # Build libamrex and all tests with CUDA 11.0.2 (recent supported)
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check -stdlib=libc++"}
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
    env:
      CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"
      CUDAFLAGS: "-stdlib=libc++"
```

File: .github/workflows/cuda.yml [hunk 6]
```diff
@@ -58,6 +58,39 @@ jobs:
         ccache -s
         du -hs ~/.cache/ccache
 
+  # Build libamrex and all tests with CUDA 11.0.2 (recent supported)
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env:
+      CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
```

File: .github/workflows/dependencies/dependencies_llvm_cuda11_ubuntu22.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+sudo apt-get update
+sudo apt-get install -y \
+    cuda-command-line-tools-11-2 \
+    cuda-compiler-11-2           \
+    cuda-cupti-dev-11-2          \
+    cuda-minimal-build-11-2      \
+    cuda-nvml-dev-11-2           \
+    cuda-nvtx-11-2               \
+    libcurand-dev-11-2
+sudo ln -s cuda-11.2 /usr/local/cuda
```
Comment (WeiqunZhang) [inline .github/workflows/dependencies/dependencies_llvm_cuda11_ubuntu22.sh]:
```suggestion
    cuda-command-line-tools-11-7 \
    cuda-compiler-11-7           \
    cuda-cupti-dev-11-7          \
    cuda-minimal-build-11-7      \
    cuda-nvml-dev-11-7           \
    cuda-nvtx-11-7               \
    libcurand-dev-11-7
sudo ln -s cuda-11.7 /usr/local/cuda
```
The oldest version of cuda on ubuntu 22.04 is 11.7.

File: .github/workflows/cuda.yml [hunk 8]
```diff
@@ -58,6 +58,55 @@ jobs:
         ccache -s
         du -hs ~/.cache/ccache
 
+  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.2
```
Comment (WeiqunZhang) [inline .github/workflows/cuda.yml]:
```suggestion
  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.7
```

**Comment by WeiqunZhang** [comment]
There is a warning.
```
clang: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
```
We can probably ignore the warning or switch to ubuntu-20.04 so that we can use cuda <= 11.5. But there is probably no apt package for clang 15 on ubuntu-20.04.

File: .github/workflows/cuda.yml [hunk 9]
```diff
@@ -58,6 +58,57 @@ jobs:
...
+        export CCACHE_COMPRESSLEVEL=10
+        export CCACHE_MAXSIZE=235M
+        ccache -z
+
+        export PATH=/usr/local/nvidia/bin:/usr/local/cuda/bin:${PATH}
+        export LD_LIBRARY_PATH=/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/cuda/lib64:${LD_LIBRARY_PATH}
+
+        cmake -S . -B build                              \
+            -DCMAKE_VERBOSE_MAKEFILE=ON                  \
+            -DAMReX_EB=ON                                \
```
Comment (WeiqunZhang) [inline .github/workflows/cuda.yml:100]:
```suggestion
            -DAMReX_MPI=OFF                              \
            -DAMReX_EB=ON                                \
```

**Comment by WeiqunZhang** [comment]
It compiles now. Ccache does not seem to work.
```
2023-06-04T22:38:36.2075501Z Cacheable calls:     0 / 223 ( 0.00%)
2023-06-04T22:38:36.2075785Z   Hits:              0
2023-06-04T22:38:36.2077008Z     Direct:          0
2023-06-04T22:38:36.2077487Z     Preprocessed:    0
2023-06-04T22:38:36.2077829Z   Misses:            0
2023-06-04T22:38:36.2078057Z Uncacheable calls: 223 / 223 (100.0%)
```

**Review by WeiqunZhang — APPROVED** [review]

**Comment by WeiqunZhang** [comment]
https://ccache.dev/releasenotes.html#_ccache_4_2
```
Added support for compiler option -x cuda, understood by Clang.
```
It should work since we use ccache 4.8.

File: .github/workflows/cuda.yml [hunk 10]
```diff
@@ -58,6 +58,58 @@ jobs:
         ccache -s
         du -hs ~/.cache/ccache
 
+  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.7
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
```
Comment (ax3l) [inline .github/workflows/cuda.yml]:
```suggestion
    name: Clang@15 CUDA@11.7 C++17 Release [tests]
```

**Comment by WeiqunZhang** [comment]
More information on why ccache fails.
```
Cacheable calls:          0 / 107 ( 0.00%)
  Hits:                   0
    Direct:               0
    Preprocessed:         0
  Misses:                 0
Uncacheable calls:      107 / 107 (100.0%)
  Preprocessing failed: 107 / 107 (100.0%)
Successful lookups:
  Direct:                 0 / 107 ( 0.00%)
  Preprocessed:           0
Local storage:
  Cache size (GB):      0.0 / 0.2 ( 0.00%)
  Files:                  0
  Hits:                   0
  Misses:                 0
  Reads:                107
  Writes:                 0
  ```

**Comment by WeiqunZhang** [comment]
Looks like the issue can be fixed with `CCACHE_DEPEND=1`. I now get
```
023-06-07T23:07:59.8206743Z Cacheable calls:    107 / 107 (100.0%)
2023-06-07T23:07:59.8207355Z   Hits:               0 / 107 ( 0.00%)
2023-06-07T23:07:59.8207876Z     Direct:           0
2023-06-07T23:07:59.8251756Z     Preprocessed:     0
2023-06-07T23:07:59.8257964Z   Misses:           107 / 107 (100.0%)
2023-06-07T23:07:59.8260542Z Successful lookups:
2023-06-07T23:07:59.8260953Z   Direct:             0 / 107 ( 0.00%)
2023-06-07T23:07:59.8261337Z   Preprocessed:       0
2023-06-07T23:07:59.8261697Z Local storage:
2023-06-07T23:07:59.8262068Z   Cache size (GB):  0.0 / 0.2 ( 5.33%)
2023-06-07T23:07:59.8262453Z   Files:            214
2023-06-07T23:07:59.8262795Z   Hits:               0 / 107 ( 0.00%)
2023-06-07T23:07:59.8263158Z   Misses:           107 / 107 (100.0%)
2023-06-07T23:07:59.8264383Z   Reads:            107
2023-06-07T23:07:59.8264638Z   Writes:           214
```
for the first run, and then
```
2023-06-07T23:11:49.2139440Z Cacheable calls:    107 / 107 (100.0%)
2023-06-07T23:11:49.2139706Z   Hits:             107 / 107 (100.0%)
2023-06-07T23:11:49.2139943Z     Direct:         107 / 107 (100.0%)
2023-06-07T23:11:49.2140206Z     Preprocessed:     0 / 107 ( 0.00%)
2023-06-07T23:11:49.2140455Z   Misses:             0 / 107 ( 0.00%)
2023-06-07T23:11:49.2140679Z Successful lookups:
2023-06-07T23:11:49.2140914Z   Direct:           107 / 107 (100.0%)
2023-06-07T23:11:49.2141152Z   Preprocessed:       0
2023-06-07T23:11:49.2141380Z Local storage:
2023-06-07T23:11:49.2141612Z   Cache size (GB):  0.0 / 0.2 ( 5.33%)
2023-06-07T23:11:49.2141830Z   Files:            214
2023-06-07T23:11:49.2142053Z   Hits:             107 / 107 (100.0%)
2023-06-07T23:11:49.2142285Z   Misses:             0 / 107 ( 0.00%)
2023-06-07T23:11:49.2142507Z   Reads:            214
2023-06-07T23:11:49.2142723Z   Writes:             0
```

**Pull request merged**